File: tests/cpp/greedy_algorithms/test_tandem_bicycle.cpp
#include "../../../problems/greedy_algorithms/tandem_bicycle/cpp/tandem_bicycle.h"
#include "../jsontestbase.h"

#include <vector>

class TandemBicycleTest : public JsonTestBase {
public:
  TandemBicycleTest() {
    json_file_path =
        std::string(TEST_CASES_DIR) + "/greedy_algorithms/tandem_bicycle.json";
  }
};

TEST_F(TandemBicycleTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty or not loeaded";

  for (const auto &test : test_cases) {
    auto redShirtSpeeds = test["redShirtSpeeds"].get<std::vector<int>>();
    auto blueShirtSpeeds = test["blueShirtSpeeds"].get<std::vector<int>>();
    bool fastest = test["fastest"].get<bool>();

    int expected = test["expected"].get<int>();
    int result = tandemBicycle(redShirtSpeeds, blueShirtSpeeds, fastest);

    EXPECT_EQ(result, expected) << "Failed for test case: " << test.dump(2);
  }
}

File: tests/python/greedy_algorithms/tandem_bicycle/test_tandem_bicycle.py
from problems.greedy_algorithms.tandem_bicycle.python.tandem_bicycle import \
    tandemBicycle
from tests.python.utility import load_test_cases


def test_tandem_bicycle() -> None:
    test_cases = load_test_cases("greedy_algorithms/tandem_bicycle.json")

    for idx, test in enumerate(test_cases):
        redShirtSpeeds: list[int] = test["redShirtSpeeds"]
        blueShirtSpeeds: list[int] = test["blueShirtSpeeds"]
        fastest: bool = test["fastest"]
        expected: int = test["expected"]

        result: int = tandemBicycle(redShirtSpeeds, blueShirtSpeeds, fastest)

        assert (
            result == expected
        ), f"Test case {idx} failed: expected {expected}, got {result}"
